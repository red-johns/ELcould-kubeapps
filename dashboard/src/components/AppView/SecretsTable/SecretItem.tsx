import * as _ from "lodash";
import * as React from "react";
import { Eye, EyeOff } from "react-feather";

import { ISecret } from "../../../shared/types";
import "./SecretContent.css";

interface ISecretItemProps {
  secret: ISecret;
}

interface ISecretItemState {
  showSecret: { [s: string]: boolean };
}

class SecretItem extends React.Component<ISecretItemProps, ISecretItemState> {
  public constructor(props: ISecretItemProps) {
    super(props);
    const showSecret = {};
    if (this.props.secret.data) {
      Object.keys(this.props.secret.data).forEach(k => (showSecret[k] = false));
    }
    this.state = { showSecret };
  }

  public render() {
    const { secret } = this.props;
    const secretEntries: JSX.Element[] = [];
    if (!_.isEmpty(this.props.secret.data)) {
      Object.keys(secret.data).forEach(k => {
        secretEntries.push(this.renderSecretEntry(k));
      });
    } else {
      secretEntries.push(<span key="empty">The secret is empty</span>);
    }
    return (
      <tr className="flex">
        <td className="col-2">{secret.metadata.name}</td>
        <td className="col-2">{secret.type}</td>
        <td className="col-7 padding-small">{secretEntries}</td>
      </tr>
    );
  }

  private renderSecretEntry = (name: string) => {
    const toggle = () => this.toggleDisplay(name);
    const text = atob(this.props.secret.data[name]);
    return (
      <span key={name} className="flex">
        <a onClick={toggle}>{this.state.showSecret[name] ? <EyeOff /> : <Eye />}</a>
        <span className="flex margin-l-normal">
          <span>{name}:</span>
          {this.state.showSecret[name] ? (
            <pre className="SecretContainer">
              <code className="SecretContent">{text}</code>
            </pre>
          ) : (
            <span className="margin-l-small">{text.length} bytes</span>
          )}
        </span>
      </span>
    );
  };

  private toggleDisplay = (name: string) => {
    const { showSecret } = this.state;
    this.setState({ showSecret: { ...showSecret, [name]: !showSecret[name] } });
  };
}

export default SecretItem;
