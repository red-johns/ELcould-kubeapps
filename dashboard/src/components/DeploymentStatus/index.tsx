import DeploymentStatus from "./DeploymentStatus";

export default DeploymentStatus;
