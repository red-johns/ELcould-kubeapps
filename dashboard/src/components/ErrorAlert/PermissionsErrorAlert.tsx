import * as React from "react";
import { Lock } from "react-feather";

import { UnexpectedErrorAlert } from ".";
import { IRBACRole } from "../../shared/types";
import { namespaceText } from "./helpers";
import PermissionsListItem from "./PermissionsListItem";

interface IPermissionsErrorPage {
  action: string;
  roles: IRBACRole[];
  namespace: string;
}

class PermissionsErrorPage extends React.Component<IPermissionsErrorPage> {
  public render() {
    const { action, namespace, roles } = this.props;
    return (
      <UnexpectedErrorAlert
        title={
          <span>
            You don't have sufficient permissions to {action} in {namespaceText(namespace)}
          </span>
        }
        icon={Lock}
        showGenericMessage={false}
      >
        <div>
          <p>Ask your administrator for the following RBAC roles:</p>
          <ul className="error__permisions-list">
            {roles.map((r, i) => (
              <PermissionsListItem key={i} namespace={namespace} role={r} />
            ))}
          </ul>
          <p>
            See the documentation for more info on{" "}
            <a
              href="https://github.com/kubeapps/kubeapps/blob/master/docs/user/access-control.md"
              target="_blank"
            >
              access control in Kubeapps
            </a>
            .
          </p>
        </div>
      </UnexpectedErrorAlert>
    );
  }
}

export default PermissionsErrorPage;
