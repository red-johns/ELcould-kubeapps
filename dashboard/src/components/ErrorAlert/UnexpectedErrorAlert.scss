.terminal__error {
  word-break: break-all;
}
