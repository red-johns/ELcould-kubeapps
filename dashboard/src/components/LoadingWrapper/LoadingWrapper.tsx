import * as React from "react";
import LoadingSpinner from "../LoadingSpinner";

export interface ILoadingWrapperProps {
  loaded?: boolean;
  size?: string;
}

// TODO(miguel) Move these kind of components to stateless compontents once we upgrade ts definitions
// Currently, I am having issues transforming it via const LoadingWrapper: React.SFC<ILoadingWrapperProps>
class LoadingWrapper extends React.Component<ILoadingWrapperProps> {
  public static defaultProps: ILoadingWrapperProps = {
    loaded: false,
  };

  public render() {
    return this.props.loaded ? this.props.children : <LoadingSpinner size={this.props.size} />;
  }
}

export default LoadingWrapper;
