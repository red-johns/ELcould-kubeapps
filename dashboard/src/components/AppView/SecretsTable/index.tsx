import SecretsTable from "./SecretsTable";

export default SecretsTable;
