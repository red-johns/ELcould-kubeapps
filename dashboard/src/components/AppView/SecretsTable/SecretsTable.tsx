import * as _ from "lodash";
import * as React from "react";

import { ErrorSelector } from "../../../components/ErrorAlert";
import LoadingWrapper from "../../../components/LoadingWrapper";
import { IKubeItem, ISecret } from "../../../shared/types";
import isSomeResourceLoading from "../helpers";
import SecretItem from "./SecretItem";

interface IServiceTableProps {
  namespace: string;
  secretNames: string[];
  secrets: Array<IKubeItem<ISecret>>;
  getSecret: (namespace: string, name: string) => void;
}

interface IError {
  resource: string;
  error: Error;
}

class SecretsTable extends React.Component<IServiceTableProps> {
  public async componentDidMount() {
    const { getSecret, secretNames, namespace } = this.props;
    secretNames.forEach(s => {
      getSecret(namespace, s);
    });
  }

  public render() {
    const { secrets } = this.props;
    return (
      <React.Fragment>
        <h6>Secrets</h6>
        <LoadingWrapper loaded={!isSomeResourceLoading(secrets)} size="small">
          {this.secretSection()}
        </LoadingWrapper>
      </React.Fragment>
    );
  }

  private findError = (): IError | null => {
    let error = null;
    _.each(this.props.secrets, (i, k) => {
      if (i.error) {
        error = { resource: k, error: i.error };
      }
    });
    return error;
  };

  private secretSection() {
    const { secrets } = this.props;
    const secretError = this.findError();
    let secretSection = <p>The current application does not contain any secret.</p>;
    if (secrets.length > 0) {
      secretSection = (
        <React.Fragment>
          <table>
            <thead>
              <tr className="flex">
                <th className="col-2">NAME</th>
                <th className="col-2">TYPE</th>
                <th className="col-7">DATA</th>
              </tr>
            </thead>
            <tbody>
              {secrets.map(
                s =>
                  s.item && <SecretItem key={`secrets/${s.item.metadata.name}`} secret={s.item} />,
              )}
            </tbody>
          </table>
          {secretError && (
            <ErrorSelector
              error={secretError.error}
              action="get"
              resource={`Secret ${secretError.resource}`}
              namespace={this.props.namespace}
            />
          )}
        </React.Fragment>
      );
    }
    return secretSection;
  }
}

export default SecretsTable;
