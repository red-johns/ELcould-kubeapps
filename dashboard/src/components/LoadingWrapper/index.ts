import LoadingWrapper, { ILoadingWrapperProps } from "./LoadingWrapper";
export { ILoadingWrapperProps };
export default LoadingWrapper;
