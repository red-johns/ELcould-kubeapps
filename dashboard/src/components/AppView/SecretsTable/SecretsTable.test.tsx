import { shallow } from "enzyme";
import context from "jest-plugin-context";
import * as React from "react";

import itBehavesLike from "../../../shared/specs";

import SecretsTable from ".";
import { ErrorSelector } from "../../../components/ErrorAlert";
import LoadingWrapper from "../../../components/LoadingWrapper";
import { ISecret } from "../../../shared/types";
import SecretItem from "./SecretItem";

context("when fetching secrets", () => {
  itBehavesLike("aLoadingComponent", {
    component: SecretsTable,
    props: {
      secretNames: [],
      secrets: [{ isFetching: true }],
    },
  });
});

context("before rendering", () => {
  it("should request the given names", () => {
    const getSecret = jest.fn();
    shallow(
      <SecretsTable namespace="foo" secretNames={["bar"]} secrets={[]} getSecret={getSecret} />,
    );
    expect(getSecret.mock.calls[0]).toEqual(["foo", "bar"]);
  });
});

const validProps = {
  namespace: "foo",
  secretNames: [],
  getSecret: jest.fn(),
  secrets: {},
};

it("renders a message if there are no secrets", () => {
  const wrapper = shallow(<SecretsTable {...validProps} secrets={[]} />);
  expect(
    wrapper
      .find(LoadingWrapper)
      .shallow()
      .find(SecretItem),
  ).not.toExist();
  expect(
    wrapper
      .find(LoadingWrapper)
      .shallow()
      .text(),
  ).toContain("The current application does not contain any secret");
});

it("renders a table with a secret", () => {
  const secrets = [
    {
      isFetching: false,
      item: {
        kind: "Secret",
        metadata: {
          name: "foo",
        },
      } as ISecret,
    },
  ];
  const wrapper = shallow(<SecretsTable {...validProps} secrets={secrets} />);
  expect(wrapper).toMatchSnapshot();
  expect(wrapper.find(SecretItem).key()).toContain("foo");
});

it("renders a table with several secrets", () => {
  const secret1 = "foo";
  const secret2 = "bar";
  const secrets = [
    { isFetching: false, item: { kind: "Secret", metadata: { name: secret1 } } as ISecret },
    { isFetching: false, item: { kind: "Secret", metadata: { name: secret2 } } as ISecret },
  ];
  const wrapper = shallow(<SecretsTable {...validProps} secrets={secrets} />);
  expect(wrapper.find(SecretItem).length).toBe(2);
  expect(
    wrapper
      .find(SecretItem)
      .at(0)
      .key(),
  ).toContain(secret1);
  expect(
    wrapper
      .find(SecretItem)
      .at(1)
      .key(),
  ).toContain(secret2);
});

it("renders a secret table with a secret and an error", () => {
  const secret1 = "foo";
  const secrets = [
    { isFetching: false, item: { kind: "Secret", metadata: { name: secret1 } } as ISecret },
    { isFetching: false, error: new Error("") },
  ];
  const wrapper = shallow(<SecretsTable {...validProps} secrets={secrets} />);
  expect(wrapper.find(SecretItem).length).toBe(1);
  expect(
    wrapper
      .find(SecretItem)
      .at(0)
      .key(),
  ).toContain(secret1);
  expect(wrapper.find(ErrorSelector)).toExist();
});
