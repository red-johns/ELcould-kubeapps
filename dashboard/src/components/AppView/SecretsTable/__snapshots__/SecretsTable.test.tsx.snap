// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a table with a secret 1`] = `
<Fragment>
  <h6>
    Secrets
  </h6>
  <LoadingWrapper
    loaded={true}
    size="small"
  >
    <table>
      <thead>
        <tr
          className="flex"
        >
          <th
            className="col-2"
          >
            NAME
          </th>
          <th
            className="col-2"
          >
            TYPE
          </th>
          <th
            className="col-7"
          >
            DATA
          </th>
        </tr>
      </thead>
      <tbody>
        <SecretItem
          key="secrets/foo"
          secret={
            Object {
              "kind": "Secret",
              "metadata": Object {
                "name": "foo",
              },
            }
          }
        />
      </tbody>
    </table>
  </LoadingWrapper>
</Fragment>
`;

exports[`when fetching secrets loading spinner matches the snapshot 1`] = `
<Fragment>
  <h6>
    Secrets
  </h6>
  <LoadingWrapper
    loaded={false}
    size="small"
  >
    <table>
      <thead>
        <tr
          className="flex"
        >
          <th
            className="col-2"
          >
            NAME
          </th>
          <th
            className="col-2"
          >
            TYPE
          </th>
          <th
            className="col-7"
          >
            DATA
          </th>
        </tr>
      </thead>
      <tbody />
    </table>
  </LoadingWrapper>
</Fragment>
`;
